db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java

```diff
@@ -26,348 +26,400 @@
+      LOG.warn("Multiple calls to 'stop()'. Scheduler is already stopping.");
+      return;
+    }
 
-			final Optional<Execution> pickedExecution = taskRepository.pick(candidate, clock.now());
+    schedulerState.setIsShuttingDown();
 
-			if (!pickedExecution.isPresent()) {
-				// someone else picked id
-				LOG.debug("Execution picked by another scheduler. Continuing to next due execution.");
-				statsRegistry.register(StatsRegistry.CandidateStatsEvent.ALREADY_PICKED);
-				return;
-			}
+    LOG.info("Shutting down Scheduler.");
+    if (!ExecutorUtils.shutdownNowAndAwaitTermination(dueExecutor, Duration.ofSeconds(5))) {
+      LOG.warn("Failed to shutdown due-executor properly.");
+    }
+    if (!ExecutorUtils.shutdownNowAndAwaitTermination(detectDeadExecutor, Duration.ofSeconds(5))) {
+      LOG.warn("Failed to shutdown detect-dead-executor properly.");
+    }
+    if (!ExecutorUtils.shutdownNowAndAwaitTermination(updateHeartbeatExecutor, Duration.ofSeconds(5))) {
+      LOG.warn("Failed to shutdown update-heartbeat-executor properly.");
+    }
 
-			currentlyProcessing.put(pickedExecution.get(), new CurrentlyExecuting(pickedExecution.get(), clock));
-			try {
-				statsRegistry.register(StatsRegistry.CandidateStatsEvent.EXECUTED);
-				executePickedExecution(pickedExecution.get());
-			} finally {
-				if (currentlyProcessing.remove(pickedExecution.get()) == null) {
-					// May happen in rare circumstances (typically concurrency tests)
-					LOG.warn("Released execution was not found in collection of executions currently being processed. Should never happen.");
-				}
-				addedDueExecutionsBatch.oneExecutionDone(() -> triggerCheckForDueExecutions());
+    if(managedExecutorService) {
+			LOG.info("Letting running executions finish. Will wait up to {}.", executorShutdownWait);
+			if (ExecutorUtils.shutdownAndAwaitTermination(executorService, executorShutdownWait)) {
+				LOG.info("Scheduler stopped.");
+			} else {
+        if(!currentlyProcessing.isEmpty()) {
+          String runningJobs = currentlyProcessing.keySet().stream().map(Execution::toString).collect(Collectors.joining("\n"));
+          LOG.warn("Scheduler did not shut down within the timeout. Interrupting the following jobs:\n{}", runningJobs);
+        } else {
+          LOG.warn("Scheduler did not shut down cleanly within the timeout. Terminating the executor.");
+        }
+			  if(!ExecutorUtils.shutdownNowAndAwaitTermination(executorService, Duration.ofSeconds(5))) {
+			    LOG.warn("Scheduler did not terminate within the timeout. The Scheduler pool may be orphaned and have live threads.");
+        }
+        if(!currentlyProcessing.isEmpty()) {
+          String staleJobs = currentlyProcessing.keySet().stream().map(Execution::toString).collect(Collectors.joining("\n"));
+          LOG.warn("Scheduler stopped, but the following jobs were orphaned:\n{}", staleJobs);
+        }
 			}
+		} else {
+    	LOG.info("Not shutting down externally managed job executor.");
```

Comment:
Indent is a bit messed up. I will probably do a reformat to 4 spaces after this PR is merged.